internal/pod/pod.go

```diff
@@ -0,0 +1,141 @@
+package pod
+
+import (
+	"context"
+	"fmt"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//go:generate mockgen -source=pod.go -package=pod -destination=mock_pod.go
+
+type ImagePuller interface {
+	CreatePullPod(ctx context.Context, imageSpec *kmmv1beta1.ModuleImageSpec, micObj *kmmv1beta1.ModuleImagesConfig) error
+	DeletePod(ctx context.Context, pod *v1.Pod) error
+	ListPullPods(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig) ([]v1.Pod, error)
+	GetPullPodForImage(pods []v1.Pod, image string) *v1.Pod
+}
+
+const (
+	moduleImageLabelKey = "kmm.node.kubernetes.io/module-image-config"
+	ImageLabelKey       = "kmm.node.kubernetes.io/module-image"
+	pullerContainerName = "puller"
+)
+
+type imagePullerImpl struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+func NewImagePuller(client client.Client, scheme *runtime.Scheme) ImagePuller {
+	return &imagePullerImpl{
+		client: client,
+		scheme: scheme,
+	}
+}
+
+func (ipi *imagePullerImpl) CreatePullPod(ctx context.Context, imageSpec *kmmv1beta1.ModuleImageSpec,
+	micObj *kmmv1beta1.ModuleImagesConfig) error {
+
+	pullPod, err := ipi.pullPodTemplate(ctx, imageSpec, micObj)
+	if err != nil {
+		return fmt.Errorf("failed to create pull pod template: %v", err)
+	}
+
+	return ipi.client.Create(ctx, pullPod)
+}
+
+func (ipi *imagePullerImpl) DeletePod(ctx context.Context, pod *v1.Pod) error {
+
+	return deletePod(ipi.client, ctx, pod)
+}
+
+func (ipi *imagePullerImpl) ListPullPods(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig) ([]v1.Pod, error) {
+
+	pl := v1.PodList{}
+
+	hl := client.HasLabels{ImageLabelKey}
+	ml := client.MatchingLabels{moduleImageLabelKey: micObj.Name}
+
+	ctrl.LoggerFrom(ctx).WithValues("mic name", micObj.Name).V(1).Info("Listing mic image Pods")
+
+	if err := ipi.client.List(ctx, &pl, client.InNamespace(micObj.Namespace), hl, ml); err != nil {
+		return nil, fmt.Errorf("could not list mic image pods for mic %s: %v", micObj.Name, err)
+	}
+
+	return pl.Items, nil
+}
+
+func (ipi *imagePullerImpl) GetPullPodForImage(pods []v1.Pod, image string) *v1.Pod {
+
+	for i, pod := range pods {
+		if image == pod.Labels[ImageLabelKey] {
+			return &pods[i]
+		}
+	}
+	return nil
+}
+
+func (ipi *imagePullerImpl) pullPodTemplate(ctx context.Context, imageSpec *kmmv1beta1.ModuleImageSpec,
```

Comment:
why do we need this function? it does not make it easier to write unit-test, and only one function is calling it. Why not make its code to be part of CreatePullPod?

Reply:
I though it is a bit simpler but it is indeed not helping unit tests and indeed no-one else is using it.
I will revert it to be how it was.

Reply:
Done.